### PR TITLE
plan: always chose the smaller child as outer for IndexJoin

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)

cherry pick https://github.com/pingcap/tidb/pull/6996 to master

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

- unit test

### CODE REVIEW

File: plan/exhaust_physical_plans.go [hunk 1]
```diff
@@ -519,22 +519,29 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *requiredProp) ([]PhysicalPlan, boo
...
 		}
 	case InnerJoin:
-		join := p.getIndexJoinByOuterIdx(prop, 0)
-		if join != nil {
-			// If the plan is not nil and matches the hint, return it directly.
-			if leftOuter {
-				return join, true
-			}
-			plans = append(plans, join...)
+		lhsCardinality := p.Children()[0].cardinality()
```
Comment (winoros) [inline plan/exhaust_physical_plans.go]: Merge master and you can use `p.Childrend()[0].StatsInfo().Count()` now.

File: plan/exhaust_physical_plans.go [hunk 2]
```diff
@@ -519,22 +519,29 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *requiredProp) ([]PhysicalPlan, boo
...
+		rhsCardinality := p.Children()[1].cardinality()
```
Comment (winoros) [inline plan/exhaust_physical_plans.go]: ditto

File: plan/plan.go [hunk 3]
```diff
@@ -184,6 +184,8 @@ type LogicalPlan interface {
 	// deriveStats derives statistic info between plans.
 	deriveStats() (*statsInfo, error)
 
+	cardinality() float64
```
Comment (winoros) [inline plan/plan.go]: We don't need this.

**Comment by zz-jason** [comment]
@winoros PTAL

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Pull request merged**